xarray/tests/test_units.py

```diff
@@ -0,0 +1,1515 @@
+    if isinstance(obj, xr.Dataset):
+        vars_units = {
+            name: array_extract_units(value) for name, value in obj.data_vars.items()
+        }
+        coords_units = {
+            name: array_extract_units(value) for name, value in obj.coords.items()
+        }
+
+        units = {**vars_units, **coords_units}
+    elif isinstance(obj, xr.DataArray):
+        vars_units = {obj.name: array_extract_units(obj)}
+        coords_units = {
+            name: array_extract_units(value) for name, value in obj.coords.items()
+        }
+
+        units = {**vars_units, **coords_units}
+    elif isinstance(obj, BaseQuantity):
+        vars_units = {"<array>": array_extract_units(obj)}
+
+        units = {**vars_units}
+    else:
+        units = {}
+
+    return units
+
+
+def strip_units(obj):
+    if isinstance(obj, xr.Dataset):
+        data_vars = {name: strip_units(value) for name, value in obj.data_vars.items()}
+        coords = {name: strip_units(value) for name, value in obj.coords.items()}
+
+        new_obj = xr.Dataset(data_vars=data_vars, coords=coords)
+    elif isinstance(obj, xr.DataArray):
+        data = array_strip_units(obj.data)
+        coords = {
+            name: (
+                (value.dims, array_strip_units(value.data))
+                if isinstance(value.data, BaseQuantity)
+                else value  # to preserve multiindexes
+            )
+            for name, value in obj.coords.items()
+        }
+
+        new_obj = xr.DataArray(name=obj.name, data=data, coords=coords, dims=obj.dims)
+    elif hasattr(obj, "magnitude"):
+        new_obj = obj.magnitude
+    else:
+        new_obj = obj
+
+    return new_obj
+
+
+def attach_units(obj, units):
+    if not isinstance(obj, (xr.DataArray, xr.Dataset)):
+        return array_attach_units(obj, units.get("data", 1))
+
+    if isinstance(obj, xr.Dataset):
+        data_vars = {
+            name: attach_units(value, units) for name, value in obj.data_vars.items()
+        }
+
+        coords = {
+            name: attach_units(value, units) for name, value in obj.coords.items()
+        }
+
+        new_obj = xr.Dataset(data_vars=data_vars, coords=coords, attrs=obj.attrs)
+    else:
+        data_units = units.get(obj.name or "data", 1)
+
+        data = array_attach_units(obj.data, data_units)
+
+        coords = {
+            name: (
+                (value.dims, array_attach_units(value, units.get(name)))
+                if name in units
+                # to preserve multiindexes
+                else value
+            )
+            for name, value in obj.coords.items()
+        }
+        dims = obj.dims
+        attrs = obj.attrs
+
+        new_obj = xr.DataArray(
+            name=obj.name, data=data, coords=coords, attrs=attrs, dims=dims
+        )
+
+    return new_obj
+
+
+def assert_equal_with_units(a, b):
+    # works like xr.testing.assert_equal, but also explicitly checks units
+    # so, it is more like assert_identical
```

Comment:
Wouldn't it be better to add pint awareness to xarray.testing.assert_identical?

Reply:
of course, that would be ideal (and I have been trying to replace it with that). But at the moment, `DataArray.identical` passes if a quantity is equal but in different units (like `10 cm` and `0.1 m`). So this is a workaround until `identical` works as expected. Also, I recently ran into an issue with converting: it returns a float, which seems to throw at least `identical` off (so maybe a combination of `allclose` and `identical`?)

Reply:
I think we should try and merge this soon so let's not get hung up. But why not add this logic into `.identical`, or at least `assert_identical`? 
i.e. `.identical` returns `False` if units are different?

Reply:
in the end having `identical` return the correct result is the goal (which we need to do anyway to have the corresponding test pass), but I would try to keep the code as general and independent from pint as possible -- though I don't know how that would look like.

To get `assert_identical` to work with quantities would require modifying `identical` and `formatting.diff_*_repr`. However, I agree that this is not urgent, so let's do that in a new PR